### PR TITLE
[Translation][FrameworkBundle] Adding Translation Providers

### LINKED ISSUE #36543: [RFC] [Translation] Remote Storages
I open this issue in collaboration with @odolbeau, who has been working on [php-translation](https://github.com/php-translation) with @nyholm, @rvanlaak, me and other contributors for a long time.

Since a few weeks we have been thinking about integrating some of [php-translation/symfony-bundle](https://github.com/php-translation/symfony-bundle) features directly into Symfony.

## Why adding new features to the Translation component?

Nowadays, the Symfony Translation component is really useful for translation process at runtime. It reads a bunch of file formats, it extracts automatically most of translation keys from your code.

However, some features are still missing. They could actually be useful for modern applications. Among these features there is the storage of translations messages in dedicated SaaS to make translators (people) independent of your application.

Moreover, the Translation Panel's DX in Symfony Profiler can be enhanced a lot.

At php-translation we have been working on those features for a few years. We think that some of those can get their right place into Symfony.

### Remote Storages

For us, the main feature to implement is Remote Storages. As I said, many modern applications use translation SaaS to manage their texts' localization.

Most of the time, those SaaS provide APIs with which developers can interact to push and pull translations.

This is very useful if your application needs to be translated by external translators. You might want to provide them a ready-to-use GUI tool and ease you to export their work once it is done. Furthermore, using these tools could avoid conflicts in translation files during development.

We would like to build Remote Storages system just as Transports in Symfony Messenger, or Adapters in Symfony Notifier. We will write an interface and let the developers write more Remote Storages if they need it.

### Configuration

Let's imagine the following configuration:

```yaml
framework:
    default_locale: en
    translator:
        default_path: '%kernel.project_dir%/translations'
        fallbacks:
            - en
        enabled: true
        logging: false
        formatter: translator.formatter.default
        cache_dir: '%kernel.cache_dir%/translations'
        paths: {  }
        
        ## Starting from here, it's new config
        remote_storages:
                # All translations in all locales are pushed and pulled to Loco
                loco:
                    dsn: loco://URL?params
                # Translations from domains "messages" & "validators" and only for locales "en" and "fr" are also pushed and pulled to transifex
                transifex:
                    dsn: transifex://URL/?params
                    domains: ['messages', 'validators']
                    locales: ['en', 'fr']
```

### Commands

We will create at least 2 commands to let you interact with translation SaaS APIs.

#### Push command

`bin/console translation:push <remote-storage-name> [--force] [--delete-obsolete] [--domains validators] [--locales en]`

This command lets you send your local translations to a Remote Storage API.
By default, we send only new translations.
With `--force` we send all translations, new included. It will overwrite updated existing remote translations.
With `--delete-obsolete` we first fetch remote translations, we filter those which do not exist anymore in local files, remove those from remote storage(s), then we send all translations, new included.

#### Pull command

`bin/console translation:pull <remote-storage-name> [--force] [--delete-obsolete] [--domains validators] [--locales en]`

This command lets you fetch your remote translations from a Remote Storage API.
By default, we fetch only new translations.
With `--force` we fetch all translations, new included. It will overwrite updated existing your local translations.
With `--delete-obsolete` simply remove your local files and recreate those from fetched remote storage(s) data. By using this one, you must be sure that your local files are up to date. We will advert you with a warning message.

> Both of these commands can filter messages on domains and/or locales.

### PR DESCRIPTION
**Pull request by welcoMattic**
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | no
| Tickets       | Fix #36543
| License       | MIT
| Doc PR        | todo

This PR refers to the early opened RFC #36543 about adding third party translation SaaS into Symfony.
I worked in collaboration with @odolbeau on this first draft.

We have implemented only Loco Provider for now, to validate the main workflow of the feature.
We are not very sure about some naming convention, such as ~~Remote Storage~~Provider, it might be renamed in Transport, to correspond to the naming of third party services in Mailer and Notifier Components.

This PR brings 2 new commands in FrameworkBundle: `translation:push` and `translation:pull`
It adds also new configuration entry:

```yaml
framework:
    default_locale: fr
    translator:
        default_path: '%kernel.project_dir%/translations'
        enabled_locales: '%locales%'
        fallbacks:
            - en
            - it
        providers:
            loco:
                dsn: '%env(LOCO_DSN)%'
                domains: ['messages']
                locales: '%locales%'
```

## To do

- [x] Implement ~~Remote Storages~~Provider into Translation component
- [x] Plug it into FrameworkBundle
- [x] Implement pull and push commands
- [x] Implement Loco adapter
- [ ] Implement Transifex Provider (:warning: API key could contains special characters like /, +, :, @, #, %, etc..., we must mention it in documentation and warn users to encode those special characters before paste them in .env)
- [x] Implement POEditor Provider (:warning: there is a trick to do in POEditor Dashboard in order to make XLF export works, it will have to be documented explicitly in the symfony/symfony-docs PR)
- [ ] Implement Crowdin Provider (:pause_button: stand by, the API is more complex to use than other Translation Management Systems)
- [ ] Tests
- [ ] Documentation
- [ ] Update CHANGELOG.md files in FrameworkBundle and Translation Component

### Transifex Provider
To do:
- [ ] Ensure that the creation of a Resource per translation key is the right way to do, if not re-work this part
- [ ] Implement API call for Translation creation
- [ ] Implement API call for Resource (means translation key) deletion

### Crowdin Provider
To do:
- [ ] Implement Add Storage and Add File API calls in order to create translations keys in Crowdin from the default locale local file
- [ ] Implement Add Translation API call to send translation messages
- [ ] Implement Update File API call in order to add or delete a translation key from a File

### CODE REVIEW

**Comment by YaFou** [comment]
**I'm really interested in this new feature! 👏** I also have another adapter which can be implemented: [Crowdin](https://crowdin.com/).

File: src/Symfony/Component/Translation/Remote/Dsn.php
Comment (Taluu) [inline src/Symfony/Component/Translation/Remote/Dsn.php:21]: So another class for a DSN... Wasn't there some discussions on that point ?

**Review by Taluu — APPROVED** [review]

File: src/Symfony/Component/Translation/Remote/Dsn.php
Comment (welcoMattic) [reply src/Symfony/Component/Translation/Remote/Dsn.php:21]: Yes, @Nyholm has opened a PR about a new DSN component (https://github.com/symfony/symfony/pull/37462)

File: src/Symfony/Component/Translation/TranslatorBagInterface.php
Comment (Pierstoval) [inline src/Symfony/Component/Translation/TranslatorBagInterface.php:30]: BC break here?

**Comment by YaFou** [comment]
@welcoMattic for the Transifex bridge, have you try to use `%2F` to escape `/` for the token?

**Comment by welcoMattic** [comment]
Before continuing the work on this PR, I ask myself several questions:

- Should we rename RemoteStorage to Transport? (like in Notifier, Mailer, ...)
- How many built-in SaaS client should we support for this first PR (should we release other "Transports" later in separate packages as it's done in Notifier?)
- Should we simplify the AbstractFactory part? (after all I think is very complex for no obvious reason)

File: src/Symfony/Bundle/FrameworkBundle/Command/TranslationPullCommand.php
Comment (fabpot) [inline src/Symfony/Bundle/FrameworkBundle/Command/TranslationPullCommand.php:29]: Should be `final class` and the phpdoc removed.

File: src/Symfony/Bundle/FrameworkBundle/Command/TranslationPullCommand.php
Comment (fabpot) [inline src/Symfony/Bundle/FrameworkBundle/Command/TranslationPullCommand.php:42]: I think `Provider` would be better than `Remote`.

File: src/Symfony/Bundle/FrameworkBundle/Command/TranslationPullCommand.php
Comment (fabpot) [inline src/Symfony/Bundle/FrameworkBundle/Command/TranslationPullCommand.php:84]: typo: remove local translation

File: src/Symfony/Bundle/FrameworkBundle/Command/TranslationPullCommand.php
Comment (fabpot) [inline src/Symfony/Bundle/FrameworkBundle/Command/TranslationPullCommand.php:92]: & -> and

File: src/Symfony/Bundle/FrameworkBundle/Command/TranslationPushCommand.php
Comment (fabpot) [inline src/Symfony/Bundle/FrameworkBundle/Command/TranslationPushCommand.php:28]: Same here `final class`

File: src/Symfony/Bundle/FrameworkBundle/Command/TranslationUpdateCommand.php
Comment (fabpot) [inline src/Symfony/Bundle/FrameworkBundle/Command/TranslationUpdateCommand.php:108]: Should be reverted. We are always using `</>` in core.

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php
Comment (fabpot) [inline src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php:1175]: `if ($config['remotes']) {`

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php
Comment (fabpot) [inline src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php:1176]: Same here, we try to never use `empty()`.

File: src/Symfony/Component/Translation/Bridge/Crowdin/CrowdinRemote.php
Comment (fabpot) [inline src/Symfony/Component/Translation/Bridge/Crowdin/CrowdinRemote.php:27]: final class

File: src/Symfony/Component/Translation/Exception/TransportException.php
Comment (fabpot) [inline src/Symfony/Component/Translation/Exception/TransportException.php:19]: all @experimental tags should be moved to 5.2.

File: src/Symfony/Component/Translation/Remote/NullRemote.php
Comment (fabpot) [inline src/Symfony/Component/Translation/Remote/NullRemote.php:21]: Is it useful?
Should be final?

File: src/Symfony/Component/Translation/Remote/RemoteDecorator.php
Comment (fabpot) [inline src/Symfony/Component/Translation/Remote/RemoteDecorator.php:16]: Why is it useful?

File: src/Symfony/Component/Translation/Remote/RemoteInterface.php
Comment (fabpot) [inline src/Symfony/Component/Translation/Remote/RemoteInterface.php:22]: Writes the given translations to the provider.

File: src/Symfony/Component/Translation/Remote/RemoteInterface.php
Comment (fabpot) [inline src/Symfony/Component/Translation/Remote/RemoteInterface.php:32]: Returns translations.

File: src/Symfony/Component/Translation/Remote/RemoteInterface.php
Comment (fabpot) [inline src/Symfony/Component/Translation/Remote/RemoteInterface.php:37]: Delete translations.

**Review by fabpot — CHANGES_REQUESTED** [review]
I've just done a very very quick review to better understand the PR and its goals.

**Comment by fabpot** [comment]
I think we would need at least 2 (or 3) providers to exercise the API and be sure it's generic and powerful enough.
I like the feature and I'd love to add it to 5.2 if possible.

**Comment by welcoMattic** [comment]
Thank you for the review. I'm ok to add 2 or 3 Providers (I will rename RemoteStorage for Provider).
I hope too include this feature in 5.2 release, I will do my best to work regularly on it, in order to be ready for 5.2.

File: src/Symfony/Component/Translation/Remote/RemoteDecorator.php
Comment (welcoMattic) [reply src/Symfony/Component/Translation/Remote/RemoteDecorator.php:16]: It is in the `read` method, that we merge wanted domains (passed as arguments in commands) with default domains (defined in configuration)

**Comment by welcoMattic** [comment]
@YaFou Hi, do you use Crowdin with Symfony? If yes, I have few questions about how Crowdin API works, especially about domains. Thanks

**Comment by YaFou** [comment]
> @YaFou Hi, do you use Crowdin with Symfony? If yes, I have few questions about how Crowdin API works, especially about domains. Thanks 

No, I'm sorry 🙄. I know some projects that use this service but I have never used it. However, if you want, I can have a look to it's API. Maybe someone else or me can add the remote after the pull request will be merged.

**Comment by welcoMattic** [comment]
@YaFou I would like to add at least 2 other Providers and Crowdin was one of those. However, we can leave it aside, and work on Phrase and POEditor (https://poeditor.com/docs/api). Those 2 looks much easier to integrate than Crowdin.
What do you think?

**Comment by andrii-bodnar** [comment]
Hi @welcoMattic, 

great work, it's a really interesting feature!

Did you consider using [Crowdin PHP client](https://github.com/crowdin/crowdin-api-client-php)? 

This is a lightweight interface to the Crowdin API v2 and can simplify interaction with Crowdin from PHP.

Also, I can assist you with any question regarding Crowdin API :slightly_smiling_face:

**Comment by welcoMattic** [comment]
Hi @andrii-bodnar! I'm still interested in implementing a Crowdin Provider here, but my main issue with Crowndin API is: how to manage Symfony's domains? I have multiple domains, and multiple locales to synchronize. Moreover, the API documentation mentions Storages, Files, Builds, which are not very clear to me. Thanks for helping us on this subject :slightly_smiling_face:

**Comment by andrii-bodnar** [comment]
@welcoMattic, could you please explain what is Symfony's domains? 

Maybe, there is some small example which can help better understand the issue

**Comment by fabpot** [comment]
To be included in Symfony, Providers (like for Mailer, Messenger, or Notifier) must not rely on external third-party packages, but should use the built-in Symfony HTTP client.

**Comment by welcoMattic** [comment]
@andrii-bodnar Domains are translations categories, like `admin`, `security`, `validators`, ... There is one file per domain and per locale (admin.en.xlf, admin.fr.xlf, admin.it.xlf, security.en.xlf, security.fr.xlf, security.if.xlf, ...). I find it hard to understand how to deal with Crowdin API model (Storage, File, Build, ...)

**Comment by welcoMattic** [comment]
@fabpot I'm working on a POEditor Provider, I try to finish it this weekend. And for the 3rd Provider I'll try to implement Crowdin one, but it doesn't seem to be going well, so I'll implement Phrase Provider in parallel, in order to make this PR almost ready for the 5.2 feature freeze

**Comment by andrii-bodnar** [comment]
@welcoMattic thanks for the explanation!

You need to upload each domain (only source language locale, e.g. **admin.en.xlf**) as a file - [Add File](https://support.crowdin.com/enterprise/api/#operation/api.projects.files.post). For this, you need to [Add Storage](https://support.crowdin.com/enterprise/api/#operation/api.storages.post). After you upload the file to the Storage, it will have a unique storage id using which you can then add the file to the project.

For locales with translations, you may use [Upload Translations](https://support.crowdin.com/enterprise/api/#operation/api.projects.translations.postOnLanguage) method. It allows to upload existing translations into the Crowdin project.

For downloading translations made in Crowdin, there is an [Export Project Translation](https://support.crowdin.com/api/v2/#operation/api.projects.translations.exports.post). This method accepts `projectId`, `targetLanguageId` and your File ID from the previous step in `fileIds` array (Also, there is more export options for different needs). 

Language ID you can get via [List Supported Languages](https://support.crowdin.com/enterprise/api/#operation/api.languages.getMany) or [Language Codes](https://support.crowdin.com/api/language-codes/)

This way you can download translations from Crowdin and put into domains with translations (e.g. **admin.fr.xlf**).

**Comment by welcoMattic** [comment]
@andrii-bodnar I found a way to do this by creating a Resource for each translation key (with i18n_type = TXT). This is what was done in https://github.com/php-translation/transifex-adapter. It seems much simpler to implement than creating a Storage (or verifying its existence), then uploading a source language file (or verifying its existence), and finally managing the translations by file. We need to manage the diff between local and remote translations finely. 
Is this way to do seems good to you, as Crowdin developer?

**Comment by andrii-bodnar** [comment]
@welcoMattic , could you please share how it looks in Crowdin?

**Comment by welcoMattic** [comment]
@andrii-bodnar It looks like:

DeepinScreenshot_select-area_20200928113006

I have one project per domain, and in each project I create a Resource per translation key.
I also find in your documentation that is possible to upload a Resource as PHP array file with many translation keys inside, but is it possible to pass this PHP array as string in the content field here: https://docs.transifex.com/api/resources#section-uploading--resources ?

**Comment by andrii-bodnar** [comment]
@welcoMattic , sorry, but it's Transifex on the screenshot, not Crowdin.

**Comment by welcoMattic** [comment]
@andrii-bodnar oh my bad sorry, I was working on both providers. I'll provide you a screenshot of my tests with Crowdin later today.

> @welcoMattic thanks for the explanation!
> 
> You need to upload each domain (only source language locale, e.g. **admin.en.xlf**) as a file - [Add File](https://support.crowdin.com/enterprise/api/#operation/api.projects.files.post). For this, you need to [Add Storage](https://support.crowdin.com/enterprise/api/#operation/api.storages.post). After you upload the file to the Storage, it will have a unique storage id using which you can then add the file to the project.
> 
> For locales with translations, you may use [Upload Translations](https://support.crowdin.com/enterprise/api/#operation/api.projects.translations.postOnLanguage) method. It allows to upload existing translations into the Crowdin project.
> 
> For downloading translations made in Crowdin, there is an [Export Project Translation](https://support.crowdin.com/api/v2/#operation/api.projects.translations.exports.post). This method accepts `projectId`, `targetLanguageId` and your File ID from the previous step in `fileIds` array (Also, there is more export options for different needs).
> 
> Language ID you can get via [List Supported Languages](https://support.crowdin.com/enterprise/api/#operation/api.languages.getMany) or [Language Codes](https://support.crowdin.com/api/language-codes/)
> 
> This way you can download translations from Crowdin and put into domains with translations (e.g. **admin.fr.xlf**).

Thanks for your explanations, I still have one more question:
Is it possible to add new translations keys without re-uploading the whole file for a domain? (e.g. I just create a new page, and I have to add new translations keys to Crowdin, or I just drop a page and I want to delete all translations keys of this page?

PS: sorry for the confusion about Crowdin/Transifex earlier.

**Comment by stof** [comment]
@welcoMattic have you looked at https://github.com/php-translation/ ? They already worked on such tooling, so the issues you have might already be solved here.

Also, this might be a good idea to have @Nyholm's opinion on this PR, as he belongs to the php-translation core team while also being part of the Symfony team (not the code core team but the recipes team).

**Comment by andrii-bodnar** [comment]
@welcoMattic,

for updating already uploaded domains to Crowdin you can use [Update or Restore File](https://support.crowdin.com/enterprise/api/#operation/api.projects.files.put) method (`storageId` here is new storage with a new domain version).

In Crowdin this file will be updated (new revision will be created) and the new keys will be added for translation (updated or deleted). Crowdin will handle all the diffs with keys within the domain.

So, the first time domain should be [added](https://support.crowdin.com/enterprise/api/#operation/api.projects.files.post) to Crowdin, and the next versions - [updated](https://support.crowdin.com/enterprise/api/#operation/api.projects.files.put).

**Comment by welcoMattic** [comment]
@stof I'm member of https://github.com/php-translation/ :wink: That's why I opened this PR. Unfortunately the only working Provider (called Adapter in php-translation) is for Loco, and we do not have Crowdin Adapter. I was inspired by the Loco Adpater of php-translation to write the Loco Provider in this PR.
Moreover, @Nyholm knows what we planned with @odolbeau about the introduction of php-translation features in Symfony, and I agree, it might be nice to have his opinion :+1:

**Comment by stof** [comment]
@welcoMattic why is there a single working adapter in php-translation ? Is it because nobody worked on others, or because they could not be implemented ? If the reason is the second case, it might indicate that the abstraction is not the good one.

**Comment by welcoMattic** [comment]
@stof it's because nobody worked on those except for Loco which I used in some projects at JoliCode.

**Comment by welcoMattic** [comment]
> @welcoMattic thanks for the explanation!
> 
> You need to upload each domain (only source language locale, e.g. **admin.en.xlf**) as a file - [Add File](https://support.crowdin.com/enterprise/api/#operation/api.projects.files.post). For this, you need to [Add Storage](https://support.crowdin.com/enterprise/api/#operation/api.storages.post). After you upload the file to the Storage, it will have a unique storage id using which you can then add the file to the project.
> 
> For locales with translations, you may use [Upload Translations](https://support.crowdin.com/enterprise/api/#operation/api.projects.translations.postOnLanguage) method. It allows to upload existing translations into the Crowdin project.
> 
> For downloading translations made in Crowdin, there is an [Export Project Translation](https://support.crowdin.com/api/v2/#operation/api.projects.translations.exports.post). This method accepts `projectId`, `targetLanguageId` and your File ID from the previous step in `fileIds` array (Also, there is more export options for different needs).
> 
> Language ID you can get via [List Supported Languages](https://support.crowdin.com/enterprise/api/#operation/api.languages.getMany) or [Language Codes](https://support.crowdin.com/api/language-codes/)
> 
> This way you can download translations from Crowdin and put into domains with translations (e.g. **admin.fr.xlf**).

@andrii-bodnar I'm trying to implement Add Storage and Add File endpoints but it's very confused for me. In the documentation, it's mentioned to send file content via Add Storage endpoint, so I send the content of my XLIFF file for the default locale. My question is: what is the purpose of the Add File endpoint as I just sent the content of my XLIFF file? It looks like a Storage represents a physical file, but what represent a File?

**Comment by andrii-bodnar** [comment]
@welcoMattic, Storage is such a "low-level" representation of a file. It's like uploading the file to our servers. This is required for any resource to be imported to Crowdin. 

After storage uploading, the file inside storage can be added as some resource (File for translation, glossary, TM, etc.)

> but what represent a File?

Files in Crowdin - is an entity containing translatable content. See [Uploading Files for Localization](https://support.crowdin.com/uploading-files/) to read more about this.

**Comment by welcoMattic** [comment]
@andrii-bodnar Hi, sorry but I do not understand how Crowdin API works for uploading translations. You mentioned the [Upload Translations endpoint](https://support.crowdin.com/enterprise/api/#operation/api.projects.translations.postOnLanguage), but in the documentation, there is no mention of content uploading. How do I upload existing translations and in which format?

**Comment by andrii-bodnar** [comment]
Hi @welcoMattic, 

let's imagine you have domain *admin.en.xlf* and French translations - *admin.fr.xlf*

File *admin.en.xlf* is being added to the Crowdin project.

Now, if you want to upload *admin.fr.xlf* translations to Crowdin, you need to create a Storage with this file and use *Upload Translations* method. This method accepts `storageId` (**admin.fr.xlf**) and `fileId` (**admin.en.xlf** file ID)

**Comment by welcoMattic** [comment]
@andrii-bodnar thank you for the quick answer. Ok, it's not very easy to understand the concepts of Storage and File, moreover with an Upload operation between them. But ok, I will try to implement this workflow in Symfony. Thanks again.

**Comment by welcoMattic** [comment]
@andrii-bodnar can we discuss by email or in [Symfony Slack](https://symfony.com/slack)? I still have difficulties to send existing translations to Crowdin.

**Comment by andrii-bodnar** [comment]
@welcoMattic, sure, I'll try to join Symfony Slack.

**Comment by stephane-lou** [comment]
This is a very welcome feature, thanks for the work guys !

**Comment by welcoMattic** [comment]
Well, there's still a lot of work here before we get to a complete and stable functionality (especially for the Provider Crowdin), and then the functional tests are not written yet.

For Crowdin, we have been working the last few days with @andrii-bodnar in order to implement all needed API calls, and it remains the `delete` method to implement, but Crowdin API does not allow fine deletion for translations keys. It's planned on their side to introduce such an endpoint very soon.

For Transifex, I need to work again the Provider, and explore the API more deeply to find the right way to do things. 

For all these reasons, I think we should target a merge for 5.3, it's more reasonable than speed up work and merge an unstable feature in 5.2.

**Comment by fabpot** [comment]
We've just moved away from `master` as the main branch to use `5.x` instead. Unfortunately, I cannot reopen the PR and change the target branch to `5.x`. Can you open a new PR referencing this one to not loose the discussion? Thank you for your understanding and for your help.

**Pull request closed without merging**
